pages/slice/language-reference/slice-grammar.md

```diff
@@ -2,14 +2,587 @@
+
+```ebnf {% showTitle=false %}
+Class
+    : Prelude "class" Identifier CompactId? (":" TypeRef)? "{" UndelimitedList<Field> "}"
+    ;
+
+CompactId
+    : "(" SignedInteger ")"
+    ;
+```
+
+For additional information on classes, see the [class][class-guide] page.
+
+### Exceptions
+
+Exceptions can only be defined or referenced in [`Slice1`][compilation-mode-guide] mode.
+Exceptions cannot be used as types. They can only validly appear in the exception specification of [operations][operation].
+
+Exception definitions consist of the `exception` keyword, followed by an [identifier], optionally followed by a base exception, and then the exception's body.
+Base exceptions are specified by a single colon, followed by an exception's (possibly scoped) identifier.
+Exception bodies consist of a list of [fields][field] wrapped in a pair of braces. These fields may be optionally separated by commas.
+Additionally, [local attributes][attribute] and a [doc-comment] may be applied to the exception in its prelude.
+
+```ebnf {% showTitle=false %}
+Exception
+    : Prelude "exception" Identifier (":" TypeRef)? "{" UndelimitedList<Field> "}"
+    ;
+```
+
+For additional information on exceptions, see the [exception][exception-guide] page.
+
+### Fields
+
+Fields can only be declared within [structs][struct], [classes][class], and [exceptions][exception].
+Field declarations consist of an [identifier], followed by a colon, and then a [type].
+Optionally, a [tag] may be applied to the field, directly before its identifier, making this a [tagged field][tag].
+Additionally, [local attributes][attribute] and a [doc-comment][doc-comment] may be applied to the field in its prelude.
+
+```ebnf {% showTitle=false %}
+Field
+    : Prelude Tag? Identifier ":" TypeRef
+    ;
+```
+
+For additional information on fields, see the [field][field-guide] page.
+
+### Interfaces
+
+Interface definitions consist of the `interface` keyword, followed by an [identifier], optionally followed by a list of base interfaces, and then the interface's body.
+Base interfaces are specified by a single colon, followed by the (possibly scoped) identifiers of one or more interfaces, separated by commas.
+Interface bodies consist of a set of [operations][operation] wrapped in a pair of braces.
+Additionally, [local attributes][attribute] and a [doc-comment] may be applied to the class in its prelude.
+
+```ebnf {% showTitle=false %}
+Interface
+    : Prelude "interface" Identifier (":" NonEmptyCommaList<TypeRef>)? "{" Operation* "}"
+    ;
+```
+
+If you are using an RPC framework, every interface definition also defines a _proxy type_.
+This proxy type shares the same name and scope as its corresponding interface and can be used as a type.
+
+```slice
+interface Foo { // Defines an interface named `Foo`.
+    op() -> Foo // Returns a proxy named `Foo`.
+}
+```
+
+It is illegal to use an interface as a type without an RPC framework.
```

Comment:
I think we should avoid the term `illegal`. Whenever I read it I always think "well its not really illegal, im not going to go to prison for it etc.". I think something more like "disallowed", "prohibited", etc. generally fits better. Illegal generally invokes the idea of the actual judicial system and law to me.

We use illegal in three spots total and this comment applies to all three.

For example:

```suggestion
Using an interface as a type is prohibited without an RPC framework.
```

Reply:
I've had the same thought before, but over the years it's left me.
'legal' and 'illegal' are pretty common in formal specifications (including things like criminal statutes),
so I don't think it's a poor fit for the reference. But I'm also not married to it.

How about:
`Interfaces cannot be used as types without an RPC framework.`? Simple and short.
Even if it's _technically_ untrue, because you _can_ do anything. It just might not compile...

Generally, I find `X is <past-tense-verb-with-ed>` out of place in specifications.
The specification isn't taking any actions, it's a passive list of statements.
I'm sure I haven't consistently stuck to that myself though :)

Reply:
The core issue here is not illegal vs cannot. This statement is incorrect: the Slice language and the Slice parser don't care and shouldn't care about RPC frameworks. 

The RPC framework aspect is purely about code generation. And that's outside the scope of the Slice language/parser.

Reply:
I removed this whole paragraph talking about proxies. I only added it because the language guide tries to draw a big line between interfaces (non-types) and proxies (types). But I always felt it was out of place here.

Now we just have interfaces and we call them types. At the language level, this is really what it looks like anyways.
We should communicate somewhere that this is not allowed unless you have an RPC framework though...